libminifi/include/utils/Id.h

```diff
@@ -65,10 +67,11 @@ class Identifier {
 
   bool operator!=(const Identifier& other) const;
   bool operator==(const Identifier& other) const;
+  bool operator<(const Identifier& other) const;
 
   bool isNil() const;
 
-  std::string to_string() const;
+  SmallString<36> to_string() const;
```

Comment:
I wonder if we could typedef this somewhere to UUIDString or something talkative, rather then just the length referring to the content.

Reply:
added UUIDString alias

Reply:
I disagree with this idea because the intent of the function is to return a string, not to return an UUIDString, the same way `std::to_string` doesn't return IntStrings and LongStrings, but just `std::string`.

In general I think these typedefs can easily obscure the code if not used with care. There doesn't seem to be an industry-wide discussion about this, so this is purely my opinion, not a guideline.

Reply:
I generally like if the types convey their usage, so we can have functions expecting `UUIDString` and we will make no mistake passing it a `"banana"` (the intent of the `Identifier::to_string` is not to return a `std::string` but to stringify its value, and that representation is quite constrained compared to `std::string`)

Reply:
on a second thought, this whole "pass identifiers as strings" interface needs a rethinking, since the only place we should call `Identifier::to_string` should be when we create logs or other outputs for human consumption

Reply:
You convinced me that returning a UUIDString is not a bad thing in itself.

> I generally like if the types convey their usage, so we can have functions expecting `UUIDString`

But they accept any `SmallString<36>` including other typedefs, the type safety is not great here and we can make the mistake of passing `"banana789012345678901234567890123456"` (36 chars). Returning or passing `UUIDString` this way is a lie because we return/accept much more than just that.

Since this is a minor detail and strong typedefs don't exist, I'm fine with either way with a preference towards `SmallString<36>` for less lying versus `UUIDString` for more talkative naming.

Reply:
As far as I see smallsring<N> cannot be constructed (even explicitly) from const char * with the proper length. Which is exactly what I would like to achieve. 
You can convert smallstring (and any typedefs based to that) to string, but you cannot anyhow create those by accident, only well designed APIs should be provide an instance of these. 

```
std::array<char, 10> temp_str = "banana123";

12:36: error: conversion from 'const char [10]' to non-scalar type 'std::array<char, 10ul>' requested
```

And I don't see any user-defined ctor in our code to make it possible. 

So the only thing we can do wrong is to typedef the same length of smallstring differently in different usecases. Which is a valid concerns as those could be mixed accidentally, but I don't think that living without typedefs helps that: if you see two APIs providing or requesting a given length of smallstr, you have no idea if they are compatible or not.

Reply:
My point is that if we follow this pattern and say we define `MySpecialMD5HashString` in the format of `"[[0123456789abcdef0123456789abcdef]]"`, with `using MySpecialMD5HashString = SmallString<36>;`, then `MySpecialMD5HashString` and `UUIDString` will be the same type and we get no errors or warnings when mixing them up.

Not a very practical example, but that's why I have a preference for avoiding such type aliases.

Reply:
Okay, I get you. 

Wonder if we can find a solution in between. For eg. UUIDString is not a typedef, but publicly inherits SmallString, constructors are private and ID (where we create those) is a friend class?

Reply:
On the flipside, that would add some boilerplate.

If we exclude the original `SmallString<36>`, I think both the current type alias and this strong typedef-like but compatible new type are acceptable.

Reply:
previously I wrote:

> I generally like if the types convey their usage, so we can have functions expecting `UUIDString`

this would make sense if we wanted functions expecting a `UUIDString`, I argue that we don't and wherever we pass/store a stringified `Identifier` we instead should pass/store an `Identifier`, [this](https://issues.apache.org/jira/browse/MINIFICPP-1395) is the issue aiming to make this change, where I think `UUIDString` would be removed

in light of this I would remove the `UUIDString` typedef

Reply:
That's indeed a better solution :+1:

Reply:
@arpadboda are you too okay with this solution?

Reply:
Works for me

Reply:
done